packages/providers/src/fetch_json.ts

```diff
@@ -24,36 +26,45 @@ export async function fetchJson(connectionInfoOrUrl: string | ConnectionInfo, js
     } else {
         connectionInfo = connectionInfoOrUrl as ConnectionInfo;
     }
+    
+    if (Array.isArray(connectionInfo.url) && !connectionInfo.url.length) {
+        throw new Error('The prioritized array of RPC Server URLs should not be empty');
+    }
 
+    const currentRpcServer = typeof connectionInfo.url === 'string' ? connectionInfo.url : connectionInfo.url[0];
     const response = await exponentialBackoff(START_WAIT_TIME_MS, RETRY_NUMBER, BACKOFF_MULTIPLIER, async () => {
         try {
 
-            const response = await (global.fetch ?? (await import('./fetch')).default)(connectionInfo.url, {
+            const response = await (global.fetch ?? (await import('./fetch')).default)(currentRpcServer, {
                 method: json ? 'POST' : 'GET',
                 body: json ? json : undefined,
-                headers: { ...connectionInfo.headers, 'Content-Type': 'application/json' }
+                headers: {
+                    'Content-Type': 'application/json',
+                    'x-api-key': connectionInfo.apiKeys ? connectionInfo.apiKeys[currentRpcServer] : undefined,
```

Comment:
I think that `null` is semantically better than `undefined` in this case
```suggestion
                    'x-api-key': connectionInfo.apiKeys ? connectionInfo.apiKeys[currentRpcServer] : null,
```